packages/components/button/src/Button.tsx

```diff
@@ -48,49 +47,83 @@ export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
     },
     ref,
   ) => {
+    const innerRef = useRef(null)
+    const buttonRef = ref || innerRef
+
     const Component = asChild ? Slot : 'button'
-    const isDisabled = !!disabled || isLoading
 
-    /**
-     * When using `asChild` (polymorphic) it's possible that the button becomes another HTML element.
-     * Depending on its tag, it could break the `disabled` html attribute preventing the clicks on a disabled button.
-     */
-    const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
-      if (isDisabled) event.preventDefault()
-      if (onClick) onClick(event)
-    }
+    const shouldNotInteract = !!disabled || isLoading
+
+    useEffect(() => {
+      /*
+       * While button is loading, we want to block interactions just as `disabled` HTML attribute would do,
+       * but still preserving focus interactions.
+       * For this reason we kind of recreate the disabled events blockage, but with some specifics.
+       */
+      if (typeof buttonRef === 'function' || !buttonRef?.current) return
+
+      const buttonElement = buttonRef.current
+
+      const blockEvent: EventListener = event => {
+        event.preventDefault()
+        event.stopPropagation()
+      }
 
-    const spinner = isLoading ? (
-      <Spinner
-        size="current"
-        className={loadingText ? 'inline-block' : 'absolute'}
-        {...(loadingLabel && { 'aria-label': loadingLabel })}
-      />
-    ) : null
+      const blockedEvents: (keyof GlobalEventHandlersEventMap)[] = [
+        'click',
+        'mousedown',
+        'mouseup',
+        'mouseenter',
+        'mouseleave',
+        'mouseover',
+        'mouseout',
+        'keydown',
+        'keypress',
+        'keyup',
+        'submit',
+      ]
+
+      if (shouldNotInteract) {
```

Comment:
I noticed that we're adding a lot of event listeners. I’m afraid this could maybe cause [performance issues](https://gomakethings.com/why-event-delegation-is-a-better-way-to-listen-for-events-in-vanilla-js/#web-performance), if there are many instances of the Button component on a page.

To avoid this, maybe we could instead create a lookup object that contains all the event handlers we need as keys, and then depending on the `shouldNotInteract` predicate, we could prevent the default event action, something like:
```tsx
// List of blocked event names that should not trigger interactions
const disabledEventsList = [
  'onClick',
  'onMouseDown',
  'onMouseUp',
  'onMouseEnter',
  'onMouseLeave',
  'onMouseOver',
  'onMouseOut',
  'onKeyDown',
  'onKeyPress',
  'onKeyUp',
  'onSubmit',
]

...

function makeDisabledEventHandlers() {
  return disabledEventsList.reduce((acc, eventName) => {
    // If no event handler was passed, skip
    if (!others[eventName]) return acc
		
    // If an event handler was passed, prevent its execution
    acc[eventName] = event => {
      event.preventDefault()
      event.stopPropagation()
    }

    return acc
  }, {})
}


...

<Component
  data-spark-component="button"
  disabled={!!disabled}
  {...others}
  {...(shouldNotInteract && makeDisabledEventHandlers())} // Override event handlers we are interested in, if those were passed
>
```

Bonus, it saves us from having to use `useEffect` and `useRef` 🧹

Reply:
You're right. If I may another suggestion, wdyt about this:
```ts
const disabledEventHandlers = useMemo(() => {
  if (!shouldNotInteract) return

  return blockedEventHandlers.reduce((acc, eventHandler) => {
      acc[eventHandler] = undefined

      return acc
    },
    {} as Record<(typeof blockedEventHandlers)[number], (() => void) | undefined>,
  )
}, [shouldNotInteract])
```

Reply:
Also, I think we can do without the `useMemo` , although I seem to be in the minority within the team, as everyone else seems to use it 😄

Reply:
Ahah :smile: Well, it seemed fitting our needs to me, to help take care of `shouldNotInteract` variable without involving to many recalculations (as for ex. in a form we may toggle this state several times)